Date Field

cvalarida opened the pull request:
# Description
Uses [`SimpleDate`](https://github.com/department-of-veterans-affairs/component-library/pull/161/) to create a date picker.

Additionally, I changed up the way the validation works so we can easily chain multiple validations together.

cvalarida commented on package.json:
```diff
@@ -24,6 +24,7 @@
     "@babel/preset-env": "^7.14.5",
     "@babel/preset-react": "^7.14.5",
     "@babel/preset-typescript": "^7.14.5",
+    "@department-of-veterans-affairs/component-library": "^2.5.7",
```
To be updated when https://github.com/department-of-veterans-affairs/component-library/pull/161/ is merged.

bkjohnson commented on src/form-builder/DateField.tsx:
```diff
@@ -0,0 +1,100 @@
+ * Super basic date string validation. Only checks for the non-empty strings
+ * between the `-`s.
+ */
+const validDate: ValidationFunction<string> = (
```
Heh, `validate`

bkjohnson commented on src/form-builder/DateField.tsx:
```diff
@@ -0,0 +1,100 @@
+
+export type ComponentLibraryDatePart = {
+  value: string;
+  dirty: boolean;
```
This does not spark joy.

cvalarida replied:
Agreed. I can probably remove it from here, honestly. With using `SimpleDate`, I don't actually _need_ it. It's never used.

bkjohnson replied:
Would removing it cause more confusion?

cvalarida replied:
Maybe not. I'll try removing it and see if stuff breaks. :sweat_smile:

bkjohnson commented on src/form-builder/DateField.tsx:
```diff
@@ -0,0 +1,100 @@
+  );
+  const id = props.id || props.name;
+
+  // TODO: Keep track of the dirty state of these
```
What would this give us?

cvalarida replied:
Yeah, I guess I'm not sure. When I wrote this, I thought it'd be useful, but... :man_shrugging: I think not. That's taken care of at the form state level, not the field level. I'll give it the :axe: !

bkjohnson commented on src/utils/validation.ts:
```diff
@@ -1,18 +1,44 @@
+  validations: ValidationFunction<T>[]
+): ((value: T) => ValidationFunctionResult<T>) => {
+  return (value: T) => {
+    // Return the error message from the first validation function that fails.
+    const errorMessage = validations
+      .map((v) => v(value, props))
+      .filter((m) => m)[0];
+    if (errorMessage) return errorMessage;
```
I feel like this behavior needs some kind of documented decision. For example, imagine a user creating a password. They receive a first error message: "Your password must have at least 18 characters". So, they add some more characters to meet the limit, then they receive a _new_ error message: "Your password must contain at least one integer". So the user makes another changes and attempts to confirm, only to be greeted with another message: "Your password must have one of these special characters: ..."

In this example I think it would be ideal to run all validations and concatenate the messages to display:

```
- Your password must have at least 18 characters
- Your password must contain at least one integer
- Your password must have one of these special characters: ...
```

In theory there might be validations which might not initially fail, but I think it may be a better experience if we show them everything that is wrong with the current value so that the user has an attempt to fix it all at once.

cvalarida replied:
That behavior is a deviation from our current UX (showing only one message at a time). To be clear, though, the behavior this function provides is such:
1. User enters a value
2. User blurs the field
    - This triggers showing the error message for the first time
    - All subsequent error messages are shown immediately using `Formik`'s default `validateOnChange={true}`, but can be changed
3. Error message is shown
4. User changes the value
5. Second error message is shown (without blurring or submitting)
6. User changes the value
7. Third error message is shown (again, while the field still has focus)

bkjohnson reviewed (commented):
Thoughts about validation and native `Date` objects.

cvalarida commented on src/typings.d.ts:
```diff
@@ -0,0 +1 @@
+declare module '@department-of-veterans-affairs/component-library/Date';
```
Yeah, so this is here to stop TypeScript from vomiting: `Cannot find type declarations for module @department-of-veterans-affairs/component-library/Date`. Ideally, we _actually_ define the types in here, but the easiest path forward is to leave the contents of the module declaration blank. The compiler will basically just assume anything defined in the module is of type `any`. See the TypeScript handbook's section on [Shorthand for Ambient Modules](https://www.typescriptlang.org/docs/handbook/modules.html#shorthand-ambient-modules) for more information.

cvalarida commented on package.json:
```diff
@@ -64,6 +65,7 @@
     "webpack-hot-middleware": "^2.25.0"
   },
   "peerDependencies": {
+    "@department-of-veterans-affairs/component-library": "^2.5.7",
```
- If a package is in `dependencies`, it shouldn't be in `peerDependencies`
- The package version in `peerDependencies` should match what's in `devDependencies`
    - We're using the version in `devDependencies` to make sure the thing works in our tests / examples
    - :point_up: means we know the downstream project needs to use at least this version

cvalarida commented on src/form-builder/DateField.tsx:
```diff
@@ -0,0 +1,100 @@
+ * component library Date component.
+ */
+const dateStringToValue = (dateString = '') => {
+  // TODO: Make this conversion handle edge cases
```
Yeah, I'm _super_ wary of using JavaScript's Date because it's...[unreliable](https://stackoverflow.com/questions/2488313/javascripts-getdate-returns-wrong-date). The short version is that it'll behave differently depending on the user's location (setting) and browser's implementation. This is part of the reason why `moment` and `date-fns` exist.

But yeah, eventually I'd like to parse this using a much more comprehensive library like `date-fns`. I'd trust that a lot more.

bkjohnson replied:
Bummer :/

cvalarida commented on src/form-builder/TextField.tsx:
```diff
@@ -35,17 +35,7 @@ const Wrapper = (props: FieldProps<string>) => {
-
-    return props.validate ? props.validate(value) : undefined;
-  };
+  const validate = chainValidations(props, [required]);
```
Are you referring to the `validate` function? In `master`, that function only adds the `required` validation. I merged in `master` and changed it up a bit so that `required` validation is now called `required` and is chainable using this `chainValidations`.

bkjohnson replied:
I think I left this comment on a bad spot - I was referring to the other non-chaining changes (try saying that three times fast) that I had...changed... in another PR.

cvalarida replied:
:joy: :+1:

bkjohnson commented on src/form-builder/DateField.tsx:
```diff
@@ -0,0 +1,99 @@
+  value: string,
+  props: DateProps
+) => {
+  if (props.required) {
+    const [month = '', day = '', year = ''] = value.split('-');
+    if (!month || !day || !year) {
+      return typeof props.required === 'string'
+        ? props.required
+        : getMessage('required.default');
+    }
+  }
```
Should `validDate` be concerned about `required`? `chainValidations` allows those validations to be treated independently, so I wonder if the individual functions should do the same.

cvalarida replied:
Maybe not. What I was thinking was that we should only verify the date is valid if the date is also required. Like, for dates that aren't required, we don't need to make sure they're complete.

But maybe that doesn't make sense after all. Without the required stipulation, it becomes "If you enter this optional date, be sure it's complete," which sounds fine. :shrug:

bkjohnson replied:
Yeah, I think it makes sense to still do some basic "is this a date" validation even if the date isn't required.

bkjohnson commented on src/form-builder/CheckboxField.tsx:
```diff
@@ -2,13 +2,16 @@ import React from 'react';
-  const withValidation = { ...props, validate: validator(props) };
+  const withValidation = {
+    ...props,
+    validate: chainValidations(props, [required]),
```
I do like that this new validation style will allow us to see each concern that we're validating for.

cvalarida commented:
Updated the validation and added a new validation message.
![image](https://user-images.githubusercontent.com/12970166/133344420-08b12c47-f8fb-46fb-a506-e66f46d645f0.png)

I used the validation message from [the old library](https://github.com/department-of-veterans-affairs/vets-website/tree/9571f54e31431939010d8e2bc2bf5a9eda02dbb8/src/platform/forms-system/src/js/validation.js).

@peggygannon Does this validation message look right to you for dates that are missing a part? I'm leaning more toward something like "Please provide a complete date," but this is what I found in the existing code.
